src/autoscaling_alloc_buffer.jl

```diff
@@ -0,0 +1,121 @@
+Additionally, the `new_buffer_size` parameter will be set to half of this size, meaning any additional allocations on the next run may increase the total memory usage by 50% (unless they are large!).
+
+This means:
+
+- `AutoscalingAllocBuffer` does not run out of memory (unlike `AllocBuffer`), since new memory will allocated on-demand when necessary
+- for repeated runs of the same size, a single contiguous buffer will be used (when `desired_max_buffers==1`), which should approximately match the performance of a tuned `AllocBuffer`
+- `AutoscalingAllocBuffer` can reuse allocated memory between runs like `AllocBuffer`, but without potentially OOMing like `SlabBuffer`. Additionally, `AutoscalingAllocBuffer` separately tracks the memory used by the main buffer vs the additional buffers allocated dynamically, so small unexpected additional allocations don't double the memory consumption (unlike a second slab being allocated).
+"""
+function AutoscalingAllocBuffer(sz::Int=DEFAULT_NEW_BUFFER_SIZE;
+                                max_history_size=DEFAULT_HISTORY_SIZE,
+                                desired_max_buffers=DEFAULT_DESIRED_MAX_BUFFERS)
+    return AutoscalingAllocBuffer(AllocBuffer(sz), [], sz, [sz], max_history_size,
+                                  desired_max_buffers)
+end
+
+# Record the size to the history without letting the history get too big
+function record_size!(v::AutoscalingAllocBuffer, new_sz::Int)
+    if length(v.main_buffer_size_history) > v.max_history_size
+        popfirst!(v.main_buffer_size_history)
+    end
+    push!(v.main_buffer_size_history, new_sz)
+    return nothing
+end
+
+function total_capacity(b::AllocBuffer)
+    return length(b.buf)
+end
+
+function amount_used(b::AllocBuffer)
+    return b.offset
+end
+
+function total_capacity(v::AutoscalingAllocBuffer)
+    return total_capacity(v.main_buffer) + sum(total_capacity, v.additional_buffers)
+end
+
+function amount_used(v::AutoscalingAllocBuffer)
+    return amount_used(v.main_buffer) + sum(amount_used, v.additional_buffers)
+end
+
+amount_left(b) = total_capacity(b) - amount_used(b)
+
+function Bumper.alloc_ptr!(v::AutoscalingAllocBuffer, sz::Int)::Ptr{Cvoid}
+    # fastest path: use main buffer if possible
+    b = v.main_buffer
+    amount_left(b) >= sz && return Bumper.alloc_ptr!(b, sz)
+    # medium path: use another available buffer
+    for b in v.additional_buffers
+        if amount_left(b) >= sz
+            return Bumper.alloc_ptr!(b, sz)
+        end
+    end
+    # slow path: allocate new memory
+    return slow_path_alloc_ptr!(v, sz)
+end
+
+@noinline function slow_path_alloc_ptr!(v::AutoscalingAllocBuffer, sz::Int)::Ptr{Cvoid}
+    old_new_buffer_size = v.new_buffer_size
+    # allocate a new buffer, defaulting to `v.new_buffer_size`, but
+    # tuning the size based on the allocatin coming in
+    new_sz = max(v.new_buffer_size, 2 * sz)
```

Comment:
Todo this should grow exponentially since otherwise if there’s a ton of small allocations we will dynamically make too many buffers 
```suggestion
    new_sz = max(2*v.new_buffer_size, 2 * sz)
```